### ISSUE TITLE
Cilium IPCache inconsistent for node IPs in cluster

### ISSUE THREAD

**Issue by joestringer** [cilium/cilium#6454]
**General Information**

- Cilium version: 1.2.6 88f6431
- Kernel version: 4.14.81-coreos

**Background**

- The cluster had previously been running v1.3.0

**Issue**

Connectivity from a pod to the local node flips between working and not working.

CIDR+L4 policy is written to allow connectivity to all node IPs, however this doesn't appear to work correctly.

**Symptoms** 

On different nodes, the local node's IP is being mapped to host, while the other identities are mapped to CIDR identities:

One node:
```
$ ks exec -ti cilium-7994f cilium map get cilium_ipcache  | grep 172.0 
172.0.112.171/32                14787   sync    
172.0.162.147/32                42668   sync    
172.0.35.235/32                 26307   sync    
172.0.163.219/32                23705   sync    
172.0.49.18/32                  17896   sync    
172.0.186.143/32                47277   sync    
172.0.39.195/32                 30143   sync    
172.0.124.60/32                 1       sync    
172.0.8.243/32                  1781    sync    
172.0.172.243/32                61548   sync    
172.0.103.94/32                 22976   sync    
172.0.88.196/32                 47616   sync    
172.0.52.103/32                 35707   sync
```

Another node:
```
$ ks exec -ti cilium-wttbz cilium map get cilium_ipcache  | grep 172.0
172.0.35.235/32                 26307   sync    
172.0.8.243/32                  1781    sync    
172.0.162.147/32                42668   sync    
172.0.172.243/32                61548   sync    
172.0.112.171/32                14787   sync    
172.0.186.143/32                47277   sync    
172.0.103.94/32                 22976   sync    
172.0.39.195/32                 30143   sync    
172.0.88.196/32                 47616   sync    
172.0.124.60/32                 34232   sync    
172.0.163.219/32                23705   sync    
172.0.52.103/32                 1       sync    
172.0.49.18/32                  17896   sync 
```

Interestingly, there is actually a CIDR identity allocated for the local node's IP, it's just not propagated into the IPCache for the local node:
```
$ ks exec -ti cilium-wttbz cilium identity list | grep 172.0
1781    cidr:172.0.8.243/32
14787   cidr:172.0.112.171/32
17896   cidr:172.0.49.18/32
22976   cidr:172.0.103.94/32
23705   cidr:172.0.163.219/32
26307   cidr:172.0.35.235/32
30143   cidr:172.0.39.195/32
34232   cidr:172.0.124.60/32
35707   cidr:172.0.52.103/32
42668   cidr:172.0.162.147/32
47277   cidr:172.0.186.143/32
47616   cidr:172.0.88.196/32
61548   cidr:172.0.172.243/32
```

**Comment by joestringer** [comment]
~~Cilium v1.3.0 is not affected by this, but has the same output of the ipcache commands above.~~

Upon upgrade to v1.3.0, the behaviour did not exhibit, however after some time this behaviour returned.

**Comment by joestringer** [comment]
For what it's worth, looking at the ipcache, all of the `cilium_host` IP addresses (typically 10.2.x.1/32 in the node's CIDR range) are being mapped to the host identity (1). However only the local node's physical (ie outward-facing) IP is considered host from an ipcache perspective. This means that traffic from local pods to node addresses may be considered using a CIDR identity or the host identity depending on where the pod is (and this is inconsistent across the cluster, with every node considering its own physical address as host).

**Comment by ianvernon** [comment]
One question I have is how we are affected during downgrade from 1.3.0 --> 1.2.6. This is because we removed the cluster identity in 1.3 (https://github.com/cilium/cilium/pull/5581). So, I believe that all nodes in the cluster are allocated their own CIDR identity if traffic is to be allowed to them by policy. However, during downgrade to v1.2.6, the cluster identity is reintroduced. But, the identities from when 1.3.0 was running are still in the key-value store, which causes confusion in the Cilium agent because the same IP is mapping to multiple identities(?). This is all conjecture, so I could very well be wrong in my assumptions.

**Comment by aanm** [comment]
@joestringer I don't fully understand what is the problem here, you were expecting `172.0.52.103/32` to have identity 1 on both nodes?

**Comment by joestringer** [comment]
@aanm If someone writes an egress CIDR policy that allows sending packets to each node in the cluster, then each node can access any other node but not the current node, even if the address of the local node is included in the CIDR policy. This is because the local node's IP is mapped to host identity.

On the flip side, if someone writes an egress policy to allow "to host", then it only allows sending traffic to the local node.

When you compare this to ingress, I believe that traffic from any node in the cluster to a pod on any node is considered as coming "from host", "to pod" identity.

**Comment by aanm** [comment]
@joestringer now I'm even more lost, I thought by default we were always allowing traffic to/from host.

**Comment by ianvernon** [comment]
@aanm IIRC we only allow ingress traffic from the host by default, not egress.

**Comment by joestringer** [comment]
@ianvernon bingo.

When the policy is working as we'd like it to, here's the policymap:

```
# cilium bpf policy get 33106 -n | grep " 1 "                                                                                                                                                                                                                                                            
Ingress     1          ANY          NONE         8233254    5592                                                                                                                                                                                                                                                               
Egress      1          ANY          NONE         3165530    6016  
```

When it's not working as intended:

```
# cilium bpf policy get 33106 -n | grep " 1 "                                                                                                                                                                                                                                                           
Ingress     1          ANY          NONE         14313798   10506
Ingress     1          9090/TCP     NONE         370331     4294
Egress      1          9404/TCP     NONE         0          0
Egress      1          9091/TCP     NONE         0          0

```

**Comment by ianvernon** [comment]
@joestringer I still don't fully grasp what exactly the traffic is that should be allowed. Is the goal to allow traffic to all hosts within the cluster on a certain port, not just the localhost?

**Comment by joestringer** [comment]
@ianvernon yes.

**Comment by ianvernon** [comment]
@joestringer ok, thanks.

The next question I have is, does this affect 1.3, or is this only an issue in downgrade from 1.3 --> 1.2.x? AFAICT this was only observed in the downgrade scenario.

**Comment by ianvernon** [comment]
@joestringer should we look into this still ? Or do you think we can close ?

**Comment by ianvernon** [comment]
Removing backport labels for the following release branches / reasons:
* v1.3 is going to be EOL'd with the release of v1.6
* v1.4 is moving to a security-fix-only backport model

I am going to nominate this for v1.6 backport so we can track this to see if it still is a problem. I will not nominate this for v1.5 backport because I'd like to reduce the churn in release branches before the most recent release (e.g., 1.6 is the most recent branch, do not backport such fixes to v1.5).
